### ISSUE TITLE
HowTo Improve Latency during echo test with TURN activated

### ISSUE THREAD

**Issue by aguerson** [bigbluebutton/bigbluebutton#10655]
Hi guys,
I hope you are fine.
I hope someone could help me.

My situation.
I use BBB 2.2.24 with 3 KMS process
I have public IP, so  I decided first to use it without STUN / TURN server.
The echo test is fast ( 1 to 3 sec )
My conf without STUN / TURN

`/etc/kurento/modules/kurento/WebRtcEndpoint.conf.ini`

```
networkInterfaces=eth0
pemCertificateRSA=/etc/kurento/dtls-srtp.pem
externalIPv4=x.x.x.1
niceAgentIceTcp=0
```

`/usr/share/bbb-web/WEB-INF/classes/spring/turn-stun-servers.xml`

```
<?xml version="1.0" encoding="UTF-8"?>
<!--

BigBlueButton open source conferencing system - http://www.bigbluebutton.org/

Copyright (c) 2012 BigBlueButton Inc. and by respective authors (see below).

This program is free software; you can redistribute it and/or modify it under the
terms of the GNU Lesser General Public License as published by the Free Software
Foundation; either version 3.0 of the License, or (at your option) any later
version.

BigBlueButton is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License along
with BigBlueButton; if not, see <http://www.gnu.org/licenses/>.

-->
<beans xmlns="http://www.springframework.org/schema/beans"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.springframework.org/schema/beans
            http://www.springframework.org/schema/beans/spring-beans-2.5.xsd
            ">

    <bean id="stun1" class="org.bigbluebutton.web.services.turn.StunServer">
        <constructor-arg index="0" value="stun:stun.freeswitch.org"/>
    </bean>

    <!--bean id="stun2" class="org.bigbluebutton.web.services.turn.StunServer">
        <constructor-arg index="0" value="stun:stun2.example.com"/>
    </bean-->

    <!--bean id="iceCandidate1" class="org.bigbluebutton.web.services.turn.RemoteIceCandidate">
        <constructor-arg index="0" value="192.168.0.1"/>
    </bean-->

    <!-- Turn servers are configured with a secret that's compatible with
         http://tools.ietf.org/html/draft-uberti-behave-turn-rest-00
         as supported by the coturn and rfc5766-turn-server turn servers -->

    <!--bean id="turn1" class="org.bigbluebutton.web.services.turn.TurnServer">
        Secret:
        <constructor-arg index="0" value="secret"/>
        TURN server URL, use turn: or turns:
        <constructor-arg index="1" value="turn:turn1.example.com"/>
        TTL in seconds for shared secret
        <constructor-arg index="2" value="86400"/>
    </bean-->

    <!--bean id="turn2" class="org.bigbluebutton.web.services.turn.TurnServer">
        <constructor-arg index="0" value="secret"/>
        <constructor-arg index="1" value="turns:turn2.example.com:443"/>
        <constructor-arg index="2" value="86400"/>
    </bean-->

    <bean id="stunTurnService" class="org.bigbluebutton.web.services.turn.StunTurnService">
        <property name="stunServers">
            <set>
                <ref bean="stun1" />
                <!--ref bean="stun2" /-->
            </set>
        </property>
        <property name="turnServers">
            <set>
                <!--ref bean="turn1" /-->
                <!--ref bean="turn2" /-->
            </set>
        </property>
        <property name="remoteIceCandidates">
            <set>
                <!--ref bean="iceCandidate1" /-->
                <!--ref bean="iceCandidate2" /-->
            </set>
        </property>
    </bean>
</beans>
```

Now I decided to use a STUN / TURN server ( coturn ) like described here : https://docs.bigbluebutton.org/2.2/setup-turn-server.html
I use a new server to install it

My new conf

`/etc/kurento/modules/kurento/WebRtcEndpoint.conf.ini`

```
networkInterfaces=eth0
stunServerAddress=x.x.x.2
stunServerPort=3478
pemCertificateRSA=/etc/kurento/dtls-srtp.pem
externalIPv4=x.x.x.1
niceAgentIceTcp=0
```

`/usr/share/bbb-web/WEB-INF/classes/spring/turn-stun-servers.xml`

```
<?xml version="1.0" encoding="UTF-8"?>
<!--

BigBlueButton open source conferencing system - http://www.bigbluebutton.org/

Copyright (c) 2012 BigBlueButton Inc. and by respective authors (see below).

This program is free software; you can redistribute it and/or modify it under the
terms of the GNU Lesser General Public License as published by the Free Software
Foundation; either version 3.0 of the License, or (at your option) any later
version.

BigBlueButton is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License along
with BigBlueButton; if not, see <http://www.gnu.org/licenses/>.

-->
<beans xmlns="http://www.springframework.org/schema/beans"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.springframework.org/schema/beans
            http://www.springframework.org/schema/beans/spring-beans-2.5.xsd">

    <bean id="stun0" class="org.bigbluebutton.web.services.turn.StunServer">
        <constructor-arg index="0" value="stun:tshosname.domain"/>
    </bean>

    <!-- Turn servers are configured with a secret that's compatible with
         http://tools.ietf.org/html/draft-uberti-behave-turn-rest-00
         as supported by the coturn and rfc5766-turn-server turn servers -->

    <bean id="turn0" class="org.bigbluebutton.web.services.turn.TurnServer">
        <constructor-arg index="0" value="yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy"/>
        <constructor-arg index="1" value="turns:tshosname.domain:443?transport=tcp"/>
        <!--TTL in seconds for shared secret-->
        <constructor-arg index="2" value="86400"/>
    </bean>

    <bean id="turn1" class="org.bigbluebutton.web.services.turn.TurnServer">
        <constructor-arg index="0" value="yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy"/>
        <constructor-arg index="1" value="turn:tshosname.domain:443?transport=tcp"/>
        <constructor-arg index="2" value="86400"/>
    </bean>

    <bean id="stunTurnService" class="org.bigbluebutton.web.services.turn.StunTurnService">
        <property name="stunServers">
            <set>
                <ref bean="stun0" />
            </set>
        </property>
        <property name="turnServers">
            <set>
                <ref bean="turn0" />
                <ref bean="turn1" />
            </set>
        </property>
        <!--
        <property name="remoteIceCandidates">
            <set>
        /-->
                <!--ref bean="iceCandidate1" /-->
                <!--ref bean="iceCandidate2" /-->
        <!--
            </set>
        </property>
        /-->
    </bean>
</beans>
```

On the coturn server

`/etc/turnserver.conf`

```
listening-port=3478
tls-listening-port=443				
listening-ip=x.x.x.2
Verbose
fingerprint
lt-cred-mech
use-auth-secret
static-auth-secret=yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy
realm=tshosname.domain
cert=/etc/ssl/certs/myownchain.pem
pkey=/etc/ssl/private/myown.key
cipher-list="ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:ECDH+3DES:DH+3DES:RSA+AESGCM:RSA+AES:RSA+3DES:!aNULL:!MD5:!DSS:!AES256"
dh2066
log-file=/var/log/coturn.log
simple-log
no-loopback-peers
no-multicast-peers
proc-user=turnserver
proc-group=turnserver
no-cli
no-tlsv1
no-tlsv1_1
```
And now, the echo test is very slow... ~15 sec
Why ?
Any solution ?

I want to pass all the trafic on the 443 port of the STUN / TRUN server, but it seems it is not possible to accelerate the echo test.

I try to put these lines in the /etc/turnserver.conf ( because I don't understand the usage )

`user=bbbuser:yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy`

and in /etc/kurento/modules/kurento/WebRtcEndpoint.conf.ini

`turnURL=bbbuser:yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy@tshosname.domain:443?transport=tls`

But nothing happened

Why I check the cortun log, I found this during the echo test

```
74: session 013000000000000001: realm <tshostname.domain> user <>: incoming packet message processed, error 401: Unauthorized
74: IPv4. Local relay addr: x.x.x.2:58743
74: session 013000000000000001: new, realm=<tshostname.domain>, username=<1602604477:w_thfpa2fj1hyk>, lifetime=3600
74: session 013000000000000001: realm <tshostname.domain> user <1602604477:w_thfpa2fj1hyk>: incoming packet ALLOCATE processed, success
74: session 010000000000000001: realm <tshostname.domain> user <>: incoming packet message processed, error 401: Unauthorized
74: IPv4. Local relay addr: x.x.x.2:56830
74: session 010000000000000001: new, realm=<tshostname.domain>, username=<1602604477:w_thfpa2fj1hyk>, lifetime=3600, cipher=ECDHE-RSA-AES128-GCM-SHA256, method=TLSv1.2
74: session 010000000000000001: realm <tshostname.domain> user <1602604477:w_thfpa2fj1hyk>: incoming packet ALLOCATE processed, success
74: session 003000000000000001: realm <tshostname.domain> user <>: incoming packet message processed, error 401: Unauthorized
74: IPv4. Local relay addr: x.x.x.2:55273
74: session 003000000000000001: new, realm=<tshostname.domain>, username=<1602604477:w_thfpa2fj1hyk>, lifetime=3600
74: session 003000000000000001: realm <tshostname.domain> user <1602604477:w_thfpa2fj1hyk>: incoming packet ALLOCATE processed, success
74: session 011000000000000001: realm <tshostname.domain> user <>: incoming packet message processed, error 401: Unauthorized
74: IPv4. Local relay addr: x.x.x.2:52332
74: session 011000000000000001: new, realm=<tshostname.domain>, username=<1602604477:w_thfpa2fj1hyk>, lifetime=3600, cipher=ECDHE-RSA-AES128-GCM-SHA256, method=TLSv1.2
74: session 011000000000000001: realm <tshostname.domain> user <1602604477:w_thfpa2fj1hyk>: incoming packet ALLOCATE processed, success
74: session 004000000000000001: realm <tshostname.domain> user <>: incoming packet message processed, error 401: Unauthorized
74: IPv4. Local relay addr: x.x.x.2:58659
74: session 004000000000000001: new, realm=<tshostname.domain>, username=<1602604477:w_thfpa2fj1hyk>, lifetime=3600
74: session 004000000000000001: realm <tshostname.domain> user <1602604477:w_thfpa2fj1hyk>: incoming packet ALLOCATE processed, success
74: session 001000000000000001: realm <tshostname.domain> user <>: incoming packet message processed, error 401: Unauthorized
74: IPv4. Local relay addr: x.x.x.2:50872
74: session 001000000000000001: new, realm=<tshostname.domain>, username=<1602604477:w_thfpa2fj1hyk>, lifetime=3600, cipher=ECDHE-RSA-AES128-GCM-SHA256, method=TLSv1.2
74: session 001000000000000001: realm <tshostname.domain> user <1602604477:w_thfpa2fj1hyk>: incoming packet ALLOCATE processed, success
74: session 009000000000000001: realm <tshostname.domain> user <>: incoming packet message processed, error 401: Unauthorized
74: IPv4. Local relay addr: x.x.x.2:64202
74: session 009000000000000001: new, realm=<tshostname.domain>, username=<1602604477:w_thfpa2fj1hyk>, lifetime=3600
74: session 009000000000000001: realm <tshostname.domain> user <1602604477:w_thfpa2fj1hyk>: incoming packet ALLOCATE processed, success
74: session 016000000000000001: realm <tshostname.domain> user <>: incoming packet message processed, error 401: Unauthorized
74: IPv4. Local relay addr: x.x.x.2:54496
74: session 016000000000000001: new, realm=<tshostname.domain>, username=<1602604477:w_thfpa2fj1hyk>, lifetime=3600, cipher=ECDHE-RSA-AES128-GCM-SHA256, method=TLSv1.2
74: session 016000000000000001: realm <tshostname.domain> user <1602604477:w_thfpa2fj1hyk>: incoming packet ALLOCATE processed, success
79: session 013000000000000001: peer x.x.x.1 lifetime updated: 300
79: session 013000000000000001: realm <tshostname.domain> user <1602604477:w_thfpa2fj1hyk>: incoming packet CREATE_PERMISSION processed, success
114: session 013000000000000001: usage: realm=<tshostname.domain>, username=<1602604477:w_thfpa2fj1hyk>, rp=1023, rb=184252, sp=1025, sb=229192
125: session 003000000000000001: refreshed, realm=<tshostname.domain>, username=<1602604477:w_thfpa2fj1hyk>, lifetime=0
```
The line 

`realm <tshostname.domain> user <>: incoming packet message processed, error 401: Unauthorized`

Is that the problem ?

Thanks in advance for the help !

**Comment by aguerson** [comment]
@ffdixon @pedrobmarin 
Hi guys,
I known you are very busy, but if you have any trick or configuration to use, I will appreciate your precious help.
Thank you again for your work !
Regards,
Aurélien.

**Comment by matiasilva** [comment]
@aguerson take a look at the existing issues here: #8824. I'm tempted to close this one as it's a duplicate of that one, is that OK?

**Comment by aguerson** [comment]
@matiasilva ;)
You are right.
I will test 
1) to upgrade to 2.2.27
2) stay to 2.2.24 ( because I can with my OpenVZ CT ) and import sip.js from https://github.com/bigbluebutton/bigbluebutton/pull/9084

I will update the issue 8824

**Comment by aguerson** [comment]
updating to 2.2.27 solved my echo problem

**Comment by webparax** [comment]
Everything was working well for me, until I updated & upgraded Ubuntu 16.04 and faced this problem!
I checked all of the answers in forums, such as IPV6 server support, udp ports being unblocked by firewall and ... .
Finally I found this guide, useful:
https://docs.bigbluebutton.org/install#1-update-your-server

But it still takes about 12 seconds after clicking on "microphone" to "echo test question" and it is annoying.
I would be so happy if anyone could guide me to reduce this period for users.
(I am sure that the server config is good: cpu 9 core = 9*2.6 GHz & Ram 16GB and I test even when there is no user and running room on the server).
